builtin/credential/jwt/path_cel_login.go

```diff
@@ -0,0 +1,329 @@
+	signingAlgorithms := toAlg(config.JWTSupportedAlgs)
+	if len(signingAlgorithms) == 0 {
+		signingAlgorithms = []jwt.Alg{
+			jwt.RS256, jwt.RS384, jwt.RS512, jwt.ES256, jwt.ES384,
+			jwt.ES512, jwt.PS256, jwt.PS384, jwt.PS512, jwt.EdDSA,
+		}
+	}
+
+	// Set expected claims values to assert on the JWT
+	expected := jwt.Expected{
+		Issuer:            config.BoundIssuer,
+		SigningAlgorithms: signingAlgorithms,
+	}
+
+	// Validate the JWT by verifying its signature and asserting expected claims values
+	allClaims, err := validator.Validate(ctx, token, expected)
+	if err != nil {
+		return logical.ErrorResponse("error validating token: %s", err.Error()), nil
+	}
+
+	// execute celRoleEntry.AuthProgram
+	role, err = b.runCelProgram(ctx, celRoleEntry, allClaims)
+	if err != nil {
+		return logical.ErrorResponse("error executing cel program: %s", err.Error()), nil
+	}
+
+	if len(role.TokenBoundCIDRs) > 0 {
+		if req.Connection == nil {
+			b.Logger().Warn("token bound CIDRs found but no connection information available for validation")
+			return nil, logical.ErrPermissionDenied
+		}
+		if !cidrutil.RemoteAddrIsOk(req.Connection.RemoteAddr, role.TokenBoundCIDRs) {
+			return nil, logical.ErrPermissionDenied
+		}
+	}
+
+	// If there are no bound audiences for the role, then the existence of any audience
+	// in the audience claim should result in an error.
+	aud, ok := getClaim(b.Logger(), allClaims, "aud").([]interface{})
+	if ok && len(aud) > 0 && len(role.BoundAudiences) == 0 {
+		return logical.ErrorResponse("audience claim found in JWT but no audiences bound to the role"), nil
+	}
+
+	alias, groupAliases, err := b.createIdentity(ctx, allClaims, celRoleEntry.Name, role, nil)
+	if err != nil {
+		return logical.ErrorResponse(err.Error()), nil
+	}
+
+	if err := validateBoundClaims(b.Logger(), role.BoundClaimsType, role.BoundClaims, allClaims); err != nil {
+		return logical.ErrorResponse("error validating claims: %s", err.Error()), nil
+	}
+
+	tokenMetadata := make(map[string]string)
+	for k, v := range alias.Metadata {
+		tokenMetadata[k] = v
+	}
+
+	auth := &logical.Auth{
+		DisplayName:  alias.Name,
+		Alias:        alias,
+		GroupAliases: groupAliases,
+		InternalData: map[string]interface{}{
+			"role": celRoleEntry.Name,
+		},
+		Metadata: tokenMetadata,
+	}
+
+	if err := role.PopulateTokenAuth(auth, req); err != nil {
+		return nil, fmt.Errorf("failed to populate auth information: %w", err)
+	}
+
+	if err := role.maybeTemplatePolicies(auth, allClaims); err != nil {
+		return nil, err
+	}
+
+	return &logical.Response{
+		Auth: auth,
+	}, nil
+}
+
+func (b *jwtAuthBackend) runCelProgram(ctx context.Context, celRoleEntry *celRoleEntry, allClaims map[string]any) (*jwtRole, error) {
+	role := jwtRole{}
+	// these functions are closures around a temporary `role` and will affect it
+	env, err := cel.NewEnv(
+		cel.Variable("claims", cel.MapType(cel.StringType, cel.DynType)),
+		cel.Function("SetPolicies",
+			cel.Overload("SetPolicies",
```

Comment:
@cipherboy I am changing this PR to evaluate CEL role login at a separate path and to use CEL functions for adjusting the resulting Auth token -- let me know if this feels like the right direction!